Objects/stringobject.c

```diff
@@ -1646,6 +1646,15 @@ string_join(PyStringObject *self, PyObject *orig)
                  * original sequence can be iterated over
                  * again, so we must pass seq here.
                  */
+                if (!PyString_Check(item) && !item->ob_bstate == NULL) {
```

Comment:
What I now realise is we probably need a `Py_GetBState(...)` function that returns (say) -1 for "not string/unicode" and then 0/1/2/3/whatever for the actual bstate. That way we can avoid putting `bstate` on *every* object.

Reply:
Today I started to add Py_GetBState, I broke other tests that I am still fixing.